### PR TITLE
Add a plugin to view Smogon tours

### PR DESCRIPTION
**Pull request by mia-pi-git**
From this: https://www.smogon.com/forums/threads/more-integration-with-smogon-tournaments.3711859
TD & US approved. Button on client to be added after this is deployed.

### CODE REVIEW

File: server/chat-plugins/smogtours.ts [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+			const [sectionID, targetID] = target.split(',').map(toID).filter(Boolean);
+			if (!sectionID || !targetID) {
+				return this.parse(`/help smogtours`);
+			}
+			const section = tours[sectionID];
+			if (!section) {
+				return this.errorReply(`Invalid section ID: "${sectionID}". Valid IDs: ${Object.keys(tours).join(', ')}`);
+			}
+			const idx = section.whitelist?.indexOf(targetID);
+			if (!section.whitelist || !idx || idx < 0) {
```
Comment (AnnikaCodes) [inline server/chat-plugins/smogtours.ts:179]:
```suggestion
			if (!idx || idx < 0) {
```

File: server/chat-plugins/smogtours.ts [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+		`Requires: &`,
+	],
+};
+
+function renderTab(inner: string, isTitle?: boolean, isCur?: boolean) {
+	isTitle = false;
+	let buf = '';
+	if (isCur) {
+		buf += `<div class="folder cur"><div class="folderhack3"><div class="folderhack1">`;
+		buf += `</div><div class="folderhack2"></div>`;
```
Comment (AnnikaCodes) [inline server/chat-plugins/smogtours.ts:210]: Why is this necessary?

File: server/chat-plugins/smogtours.ts [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+		buf += `<div class="selectFolder">${inner}</div></div></div>`;
+	} else {
+		if (!isTitle) {
+			inner = `<div class="selectFolder">${inner}</div>`;
```
Comment (AnnikaCodes) [inline server/chat-plugins/smogtours.ts:214]: you should add a jsdoc comment to this function to explain that it mutates `inner`

File: server/chat-plugins/smogtours.ts [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+ */
+import {FS, Utils} from '../../lib';
+
+interface TourEvent {
+	title: string;
+	url: string;
+	desc: string;
+	image?: {url: string, dimensions: [number, number]};
+	id: string;
+	shortdesc: string;
```
Comment (KrisXV) [inline server/chat-plugins/smogtours.ts:13]:
```suggestion
	shortDesc: string;
```
im a camelcase loser

File: server/chat-plugins/smogtours.ts [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+
+function saveTours() {
+	FS('config/chat-plugins/smogtours.json').writeUpdate(() => JSON.stringify(tours));
+}
+
+function getTour(category: string, id: string) {
+	category = toID(category);
+	id = toID(id);
+	return tours?.[category].tours.find(f => f.id === id) || null;
+}
```
Comment (KrisXV) [inline server/chat-plugins/smogtours.ts:64]:
```suggestion
function getTour(categoryId: ID, tournamentName: string) {
	const tournamentId = toID(tournamentName);
	return tours?.[categoryId].tours.find(f => f.id === tournamentId) || null;
}
```
I think forcing the category ID out of the function ensures more type safety, but nonetheless I feel like more variable name clarity is useful

File: server/chat-plugins/smogtours.ts [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+			this.checkCan('rangeban');
+			const [sectionID, targetID] = target.split(',').map(toID).filter(Boolean);
+			if (!sectionID || !targetID) {
+				return this.parse(`/help smogtours`);
+			}
+			const section = tours[sectionID];
+			if (!section) {
+				return this.errorReply(`Invalid section ID: "${sectionID}". Valid IDs: ${Object.keys(tours).join(', ')}`);
+			}
+			const idx = section.whitelist?.indexOf(targetID);
```
Comment (KrisXV) [inline server/chat-plugins/smogtours.ts:178]:
```suggestion
			const index = section.whitelist?.indexOf(targetID);
```
nitpick but i feel like reducing a 5 letter variable name to 3 isnt necessary when there isnt a higher level variable w/ the same name

File: server/chat-plugins/smogtours.ts [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+		`/smogtours unwhitelist [section], [user] - Removes the given [user] from the [section]'s management whitelist.`,
+		`Requires: &`,
+	],
+};
+
+function renderTab(inner: string, isTitle?: boolean, isCur?: boolean) {
+	isTitle = false;
+	let buf = '';
+	if (isCur) {
+		buf += `<div class="folder cur"><div class="folderhack3"><div class="folderhack1">`;
```
Comment (KrisXV) [inline server/chat-plugins/smogtours.ts:209]: Potentially too much but would it be possible to make this a `.tsx` file :3

File: server/chat-plugins/smogtours.ts [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+	}
+	return buf;
+}
+
+const refresh = (pageid: string) => (
+	`<button class="button" name="send" value="/join ${pageid}" style="float: right">` +
+	`<i class="fa fa-refresh"></i> Refresh</button>`
+);
+
+export function renderPageChooser(curPage: string, buffer: string, user?: User) {
```
Comment (KrisXV) [inline server/chat-plugins/smogtours.ts:226]: Is there a reason to export this?

**Comment by KrisXV** [comment]
Potentially making a class full of the helper functions so there's _some_ cleanliness might help instead of having them laid out at different places in the file?

File: server/chat-plugins/smogtours.ts (hunk 2 above)
Comment (mia-pi-git) [reply server/chat-plugins/smogtours.ts:210]: The CSS absolutely breaks otherwise. No idea why.

File: server/chat-plugins/smogtours.ts (hunk 8 above)
Comment (mia-pi-git) [reply server/chat-plugins/smogtours.ts:226]: It was entirely for /eval testing. No other purpose otherwise.

File: server/chat-plugins/smogtours.ts [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+	image?: {url: string, dimensions: [number, number]};
+	id: string;
+	shortdesc: string;
+}
+
+interface TourTable {
+	title: string;
+	tours: TourEvent[];
+	whitelist?: string[];
+	icon?: [string, number, number];
```
[the remark this answers is not in the document]
Comment (mia-pi-git) [reply server/chat-plugins/smogtours.ts:20]: I didn't because I kept them separate in `TourEvent`. Suppose there's no reason not to.

**Comment by mia-pi-git** [comment]
> Potentially making a class full of the helper functions so there's _some_ cleanliness might help instead of having them laid out at different places in the file?

I'm not super against this but I would like to note I already did order everything fairly reasonably - one group of functions is all the HTML helpers, and the other is permissions related.

File: server/chat-plugins/smogtours.ts (hunk 1 above)
Comment (mia-pi-git) [reply server/chat-plugins/smogtours.ts:178]: This is necessary to ensure TypeScript doesn't complain when it's referenced further down.

File: server/chat-plugins/smogtours.ts (hunk 2 above)
Comment (AnnikaCodes) [reply server/chat-plugins/smogtours.ts:228]: Can you at least add a comment saying so?

File: server/chat-plugins/smogtours.ts [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+		title: "Smogon Officials",
+		// cap this one's dimensions
+		icon: ['https://www.smogon.com/media/zracknel-beta.svg', 178, 200],
+		tours: [],
+		desc: "Tournaments run by Smogon staff.",
+	},
+	smogon: {
+		title: "Open Sign-Ups",
+		tours: [],
+		desc: "Tournaments run by Smogon staff and regular users alike."
```
Comment (AnnikaCodes) [inline server/chat-plugins/smogtours.ts:36]: ci failure

File: server/chat-plugins/smogtours.ts (hunk 2 above)
Comment (mia-pi-git) [reply server/chat-plugins/smogtours.ts:228]: OK.

File: server/chat-plugins/smogtours.ts (hunk 10 above)
Comment (mia-pi-git) [reply server/chat-plugins/smogtours.ts]: Fixed.

File: server/chat-plugins/smogtours.ts [hunk 11]
```diff
@@ -0,0 +1,447 @@
...
+			// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/test
+			Chat.linkRegex.lastIndex = -1;
+			if (!Chat.linkRegex.test(url)) {
+				return this.popupReply(`Invalid info URL: "${url}"`);
+			}
+			let image;
+			if (rawImg) {
+				// see above
+				Chat.linkRegex.lastIndex = -1;
+				if (!Chat.linkRegex.test(rawImg)) {
```
Comment (AnnikaCodes) [inline server/chat-plugins/smogtours.ts:119]: you might want to make a test function that sets `lastIndex` appropriately before using the regex

File: server/chat-plugins/smogtours.ts [hunk 12]
```diff
@@ -0,0 +1,447 @@
...
+				`${user.name} removed ${targetID} from the tour management whitelist for the ${section.title} section`
+			);
+			this.globalModlog('TOUR UNWHITELIST', targetID);
+			saveTours();
+		},
+		view() {
+			return this.parse(`/join view-tournaments-all`);
+		},
+	},
+	smogtourshelp: [
```
Comment (AnnikaCodes) [inline server/chat-plugins/smogtours.ts:213]: "smogtours" makes me think of smogtours.psim.us; maybe "forum tours" would be a better name?

File: server/chat-plugins/smogtours.ts [hunk 13]
```diff
@@ -0,0 +1,447 @@
...
+		`/smogtours view - View a list of ongoing forum tournaments.`,
+		`/smogtours whitelist [section], [user] - Whitelists the given [user] to manage tournaments for the given [section].`,
+		`Requires: &`,
+		`/smogtours unwhitelist [section], [user] - Removes the given [user] from the [section]'s management whitelist.`,
+		`Requires: &`,
+	],
+};
+
+/** Wrapps `inner` in the necessary HTML to show a tab on the sidebar */
```
Comment (AnnikaCodes) [inline server/chat-plugins/smogtours.ts:222]:
```suggestion
/** Modifies `inner` in-place to wrap it in the necessary HTML to show a tab on the sidebar. */
```

File: server/chat-plugins/smogtours.ts (hunk 12 above)
Comment (mia-pi-git) [reply server/chat-plugins/smogtours.ts:213]: I would prefer not - this is not meant for _exclusively_ forums.

File: server/chat-plugins/smogtours.ts (hunk 11 above)
Comment (mia-pi-git) [reply server/chat-plugins/smogtours.ts]: ok

File: server/chat-plugins/smogtours.ts [hunk 14]
```diff
@@ -0,0 +1,448 @@
...
+			}
+			const section = tours[sectionID];
+			if (!section) {
+				return this.errorReply(`Invalid section ID: "${sectionID}". Valid IDs: ${Object.keys(tours).join(', ')}`);
+			}
+			const idx = section.whitelist?.indexOf(targetID);
+			if (!section.whitelist || !idx || idx < 0) {
+				return this.errorReply(`${targetID} is not whitelisted in that section.`);
+			}
+			section.whitelist?.splice(idx, 1);
```
Comment (MathyFurret) [inline server/chat-plugins/smogtours.ts:199]:
```suggestion
			section.whitelist.splice(idx, 1);
```

**Review by DaWoblefet — APPROVED** [review]
I did extensive functionality review with Mia and all concerns have been addressed. Looks great to me from that perspective!
